src/bin/cargo/cli.rs

```diff
@@ -222,16 +219,30 @@ fn add_ssl(version_string: &mut String) {
     }
 }
 
+/// Expands aliases recursively to collect all the command line arguments.
+///
+/// [`GlobalArgs`] need to be extracted before expanding aliases because the
+/// clap code for extracting a subcommand discards global options
+/// (appearing before the subcommand).
 fn expand_aliases(
     config: &mut Config,
     args: ArgMatches,
     mut already_expanded: Vec<String>,
 ) -> Result<(ArgMatches, GlobalArgs), CliError> {
     if let Some((cmd, args)) = args.subcommand() {
-        match (
-            commands::builtin_exec(cmd),
-            super::aliased_command(config, cmd)?,
-        ) {
+        let exec = commands::builtin_exec(cmd);
+        let aliased_cmd = super::aliased_command(config, cmd).or_else(|e| {
+            // HACK: `cargo version` must always work even with malformed configs.
+            // Here we treat the error as the aliased command not found,
+            // so it falls back to call builtin `cargo version`
+            if cmd == "version" {
+                Ok(None)
+            } else {
+                Err(e)
+            }
+        })?;
+
+        match (exec, aliased_cmd) {
```

Comment:
This is making us lose out on the warning when a user aiases a built-in

Reply:
As far as I understanding it, the config file is already broken, so no alias can be resolved. We don't lose any information here because this is a recovery from an error.

Reply:
Right, its in an `or_else` so its already failed.